### ISSUE TITLE
Updating settings allows arbitrary values

### ISSUE THREAD

**Issue by dometto** [repotag/repotag#10]
At the moment it seems possible to assign arbitrary values to a key in a settings hash. For example, you could set an "enabled" setting to a string by passing `:enabled => "somestring"` into the `params` hash. Or you could set any key to nil by passing `:key => ""`. But of course in the app we want this key to pertain to a Boolean.

Is this a problem? If so, is there a way to be strict about the types of values you can set for a given key? Using strong parameters we can declare that certain kinds of values are permitted, but the tricky part will be deciding _which_ kind of values should be allowed for which keys.

**Comment by dometto** [comment]
We could give each Setting an (optional) second hash attribute which would define allowed attributes and values? Then we could set them [here](https://github.com/repotag/repotag/blob/master/config/initializers/app_defaults.rb), and do a validation in the `Setting#[]=` method?

**Comment by bartkamphorst** [comment]
We could, but that would seem like a Ruby anti-pattern to me. What problems do you foresee?

**Comment by dometto** [comment]
Yeah, I don't like the feel of it either. But I also don't like the idea that you could e.g. set an intended boolean field to some arbitrary string. How do we (uniformly) handle casting that to a boolean?

Another possible problem is with setting values to nil: how do we tell the difference between a field that the user has set to nil, and a field that just has not been set? For instance, if the SMTP server host setting is nil, do we assume the user doesn't want mail to be sent at all, or that we should fallback to the default value (`localhost`)?

Maybe we don't need to handle these edge cases by strict type checking, but I think we should have some kind of strategy for them -- I have the feeling such things will lead to bugs in the future if kept unchecked.

**Comment by bartkamphorst** [comment]
Okay, I've now encountered these cases. :smile: In particular, the `:enable_starttls_auto` setting requires a boolean, where xeditable sets either a `'1'` or a `'0'`.

**Comment by dometto** [comment]
Can we provide Setting#get with a way to transform reserved values like "1" and "0" into certain other values?

**Comment by dometto** [comment]
One idea is to specify a custom class for handling serialization which does the required conversions either when reading from or when writing to the database. See http://apidock.com/rails/ActiveRecord/AttributeMethods/Serialization/ClassMethods/serialize

**Comment by bartkamphorst** [comment]
I'm not entirely sure how that would help. Because I would like `smtp_settings[:enable_starttls_auto]` to be an actual boolean when I'm evaluating it, but a '1' or '0' for xeditable. I think we should instead look for a way of [improving the way xeditable handles things](https://github.com/vitalets/x-editable/issues/138).

**Comment by dometto** [comment]
That would help to make sure string values for booleans arent passed from the webinterface. It would still theoretically be possible for a string value to end up in an intended boolean field though. So should we just check whether the value is truthy everwhere in the app and accept that the db might contain strings instead of booleans?

By the way, another potential worry is that a user could set settings values to extremely long strings. In the db each string field has a max length of course, but I'm not sure how this works for serialized hashes stored in the db.

**Comment by dometto** [comment]
This gem actually supports settings with validations:

https://github.com/ledermann/rails-settings
